### PR TITLE
Move doc sections to "guide" + "reference"

### PR DESCRIPTION
**Pull request by mcabbott**
This is one take on how we might aim to organise the docs. No new text, just moving headings around. 

The goal is to separate it into
* A guide you could read straight through, and encounter each idea once. Basics first, more specialised later.
* A reference section, to look up all the weird loss functions / layers / etc, mostly just docstrings
* A tutorials section, into which we move what's now on the website, and maybe the model zoo. Each should solve some problem, ideally data -> model -> results. Fine if they overlap. 

This seems like a non-awful way to present things, which can be achieved without re-writing more than we have to. The sections which most awkwardly straddle guide/reference are the training/optimisation/regularisation ones, which need re-working for explicit params anyway. #2114 is a start. 

One downside of this is that some things like `onehot` and `outputsize` don't really appear in the "guide" half at all. Maybe they ought to be squeezed in somewhere? 

### Edit, more comments:

~~Maybe "Fitting a Line" and "Quickstart" also straddle the guide/tutorial boundary a bit. Could think about moving them there.~~ Maybe I like the choice of fast/slow intro. I see now that this was what the original overview PR envisaged: https://github.com/FluxML/Flux.jl/pull/1579#issue-860485363 (naming exactly xor as what the quickstart should target!)

Maybe the guide pages should be consolidated a bit -- "performance tips" doesn't say much, can it be rolled into "GPU & performance tips"? "Regularisation" also doesn't say much ... #2114  wants to make it just a section.

Not sure what order these sections ought to be in. One complaint in https://github.com/FluxML/Flux.jl/issues/2105 is "Ecosystem...  is currently way down at the bottom". This could move up, although we can't put everything first... what exactly deserves to go last? 

Do tutorials come after reference? Perhaps #2125 was merged in haste, as these tutorials aren't actually much good... maybe one or two should be updated & the rest trashed. The versions on the model zoo have stayed more up to date, maybe people should just read those -- the added text on these versions is often more confusing than helpful.

### CODE REVIEW

**Comment by ToucheSir** [comment]
Is there really no way to selectively collapse larger sidebar sections in Documenter by default? I feel like that would make the high-level order much less important and save us some decision fatigue.

**Comment by mcabbott** [comment]
I looked at a few packages linked from Documenter's docs, and don't see any such thing. But I don't think it's too bad really, you don't have to scroll so far that learning how to use some folding thing would pay.

I wonder today if we should make Ecosystem a top-level heading, with a page for packages containing Flux models, separate from a page of less closely associated links. 

I also wonder if we want a page of CUDA docstrings, separate from the current GPU page. For now, https://fluxml.ai/Flux.jl/previews/PR2115/CUDA/

File: docs/make.jl [hunk 1]
```diff
@@ -1,59 +1,62 @@
...
+        # Aim is to cover each new concept exactly once (but not list all variants).
+        # Hard to invent further divisions which aren't more confusing than helpful?
             "Quick Start" => "models/quickstart.md",
             "Fitting a Line" => "models/overview.md",
             "Gradients and Layers" => "models/basics.md",
-        ],
-        "Building Models" => [
-            "Built-in Layers 📚" => "models/layers.md",
+            "Training" => "training/training.md",
+            # "Regularisation" => "models/regularisation.md",  # consolidated in #2114
```
Comment (mcabbott) [inline docs/make.jl:21]: PR started out just re-organising sections. 

This commented line should probably stay until #2114

File: docs/src/ecosystem.md [hunk 2]
```diff
@@ -104,8 +108,22 @@ Some useful and random packages!
...
+- [Parameters.jl](https://github.com/mauro3/Parameters.jl) types with default field values, keyword constructors and (un-)pack macros.
 - [BSON.jl](https://github.com/JuliaIO/BSON.jl) is a package for working with the Binary JSON serialisation format.
 - [DataFrames.jl](https://github.com/JuliaData/DataFrames.jl) in-memory tabular data in Julia.
 - [DrWatson.jl](https://github.com/JuliaDynamics/DrWatson.jl) is a scientific project assistant software.
 
 This tight integration among Julia packages is shown in some of the examples in the [model-zoo](https://github.com/FluxML/model-zoo) repository.
+
+---
+
+## Alternatives to Flux
```
Comment (mcabbott) [inline docs/src/ecosystem.md:120]: Small changes to the links page. The biggest is this section on SimpleChains / KNet / Lux. See what you think?

File: docs/src/index.md [hunk 3]
```diff
@@ -2,27 +2,23 @@
...
+* **Doing the obvious thing**. Flux has relatively few explicit APIs. Instead, writing down the mathematical form will work – and be fast.
+* **Extensible by default**. Flux is written to be highly flexible while being performant. Extending Flux is as simple as using your own code as part of the model you want - it is all [high-level Julia code](https://github.com/FluxML/Flux.jl/tree/master/src).
+* **Play nicely with others**. Flux works well with unrelated Julia libraries from [images](https://github.com/JuliaImages/Images.jl) to [differential equation solvers](https://github.com/SciML/DifferentialEquations.jl), rather than duplicating them.
 
 ## Installation
 
-Download [Julia 1.6](https://julialang.org/downloads/) or later, preferably the current stable release. You can add Flux using Julia's package manager, by typing `] add Flux` in the Julia prompt.
-
-This will automatically install several other packages, including [CUDA.jl](https://github.com/JuliaGPU/CUDA.jl) which supports Nvidia GPUs. To directly access some of its functionality, you may want to add `] add CUDA` too. The page on [GPU support](gpu.md) has more details.
```
Comment (mcabbott) [inline docs/src/index.md:13]: I tried to delete lots of text from the welcome page. It was pretty verbose.

File: docs/src/models/basics.md [hunk 4]
```diff
@@ -232,6 +232,5 @@ Affine(3 => 1, bias=false, init=ones) |> gpu
 ```
 
 ```@docs
-Functors.@functor
```
Comment (mcabbott) [inline docs/src/models/basics.md:235]: This macro has its (long) docstring in the Functors page, so it need not appear here (and in fact Documenter complains about it appearing twice)

File: docs/src/reference/CUDA.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
+# CUDA.jl
+
+I'm not entirely sure this page should be separate from the gpu.md page.
```
Comment (mcabbott) [inline docs/src/reference/CUDA.md:3]: Should this page exist? So far all the "other package" pages are for closely associated packages. Maybe we should just link.

Comment (ToucheSir) [reply docs/src/reference/CUDA.md]: I like the link idea too. Or multiple links to specific pages like the CUDA.jl benchmarking tips.

Comment (mcabbott) [reply docs/src/reference/CUDA.md]: I deleted this big page, and put a few docstrings at https://fluxml.ai/Flux.jl/previews/PR2115/gpu/#CUDA-Reference  to see how they look. But perhaps too many, at least for this PR.

Comment (ToucheSir) [reply docs/src/reference/CUDA.md]: I saw those, this seemed like the place to comment though. If a link saves us from having to deal with CUDA.jl as a test dep, that's a worthy tradeoff for me.

Comment (mcabbott) [reply docs/src/reference/CUDA.md]: OK I've removed it completely. The GPU page could use some tidying but not this PR.

File: docs/src/ecosystem.md [hunk 6]
```diff
@@ -104,8 +108,22 @@ Some useful and random packages!
...
+
+Julia has several other libraries for making neural networks. 
+
+* [SimpleChains.jl](https://github.com/PumasAI/SimpleChains.jl) is focused on making small, simple, CPU-based, neural networks fast. Uses [LoopVectorization.jl](https://github.com/JuliaSIMD/LoopVectorization.jl). (Was `FastChain` in DiffEqFlux.jl) 
+
+* [Knet.jl](https://github.com/denizyuret/Knet.jl) is a neural network library built around [AutoGrad.jl](https://github.com/denizyuret/AutoGrad.jl), with beautiful documentation.
```
Comment (ToucheSir) [inline docs/src/ecosystem.md:126]:
```suggestion
* [Knet.jl](https://github.com/denizyuret/Knet.jl) is a neural network library built around [AutoGrad.jl](https://github.com/denizyuret/AutoGrad.jl).
```
Is "with beautiful documentation" from a tagline somewhere? Trying to think of a better point of comparison for this bullet.

File: docs/src/index.md [hunk 7]
```diff
@@ -2,29 +2,25 @@
...
+* **Extensible by default**. Flux is written to be highly flexible while being performant. Extending Flux is as simple as using your own code as part of the model you want - it is all [high-level Julia code](https://github.com/FluxML/Flux.jl/tree/master/src).
+* **Play nicely with others**. Flux works well with unrelated Julia libraries from [images](https://github.com/JuliaImages/Images.jl) to [differential equation solvers](https://github.com/SciML/DifferentialEquations.jl), rather than duplicating them.
 
-## Installation
+### Installation
 
-Download [Julia 1.6](https://julialang.org/downloads/) or later, preferably the current stable release. You can add Flux using Julia's package manager, by typing `] add Flux` in the Julia prompt.
+Download [Julia 1.6](https://julialang.org/downloads/) or later, preferably the current stable release. You can add Flux using Julia's package manager, by typing `] add Flux` in the Julia prompt. This will automatically install several other packages, including [CUDA.jl](https://github.com/JuliaGPU/CUDA.jl) which supports Nvidia GPUs.
```
Comment (ToucheSir) [inline docs/src/index.md:11]:
```suggestion
Download [Julia 1.6](https://julialang.org/downloads/) or later, preferably the current stable release. You can add Flux using Julia's package manager, by typing `] add Flux` in the Julia prompt. This will automatically install several other packages, including [CUDA.jl](https://github.com/JuliaGPU/CUDA.jl) for Nvidia GPU support.
```

File: docs/make.jl [hunk 8]
```diff
@@ -9,51 +9,53 @@ makedocs(
...
+            "Loss Functions" => "models/losses.md",
+            "Optimisation Rules" => "training/optimisers.md",  # TODO move optimiser intro up to Training
+            "Shape Inference" => "outputsize.md",
+            "Flat vs. Nested" => "destructure.md",
+            "Callback Helpers" => "training/callbacks.md",
+            "NNlib.jl (`softmax`, `conv`, ...)" => "models/nnlib.md",
+            "Zygote.jl (`gradient`, ...)" => "training/zygote.md",
+            "MLUtils.jl (`DataLoader`, ...)" => "data/mlutils.md",
+            "Functors.jl (`fmap`, ...)" => "models/functors.md",
+            "OneHotArrays.jl (`onehot`, ...)" => "data/onehot.md",
```
Comment (ToucheSir) [inline docs/make.jl:42]: Having example functions right in the heading feels a little strange. Could we instead use adapt the tagline of each page? e.g. "Neural Network primitives - NNlib.jl".

File: docs/src/ecosystem.md (hunk 6 above)
Comment (mcabbott) [reply docs/src/ecosystem.md:126]: It is really nice, https://denizyuret.github.io/Knet.jl/latest/mlp/

But I agree this may read as a dismissal. The entire section is a bit of a minefield...

File: docs/make.jl (hunk 8 above)
Comment (mcabbott) [reply docs/make.jl:42]: I don't love them, but they are compact, and I didn't manage to think of a better way (last time around). I mean I know what "primitives" means but it's not so obvious... what kind of utils is MLUtils, what on earth is Functors in less than a paragraph?

File: docs/src/index.md [hunk 9]
```diff
@@ -2,29 +2,25 @@
...
 
-The [quick start](@ref man-quickstart) page trains a simple neural network.
+There are some **tutorials** about building particular models. The **[model zoo](https://github.com/FluxML/model-zoo/)** has starting points for many other common ones. And finally, the **[ecosystem page](ecosystem.md)** lists packages which define Flux models.
 
-This rest of this documentation provides a from-scratch introduction to Flux's take on models and how they work, starting with [fitting a line](@ref man-overview). Once you understand these docs, congratulations, you also understand [Flux's source code](https://github.com/FluxML/Flux.jl), which is intended to be concise, legible and a good reference for more advanced concepts.
+The **reference** section contains API listings, including some companion packages: [Zygote](https://github.com/FluxML/Zygote.jl) (automatic differentiation), [Optimisers](https://github.com/FluxML/Optimisers.jl) (training), [NNlib](https://github.com/FluxML/NNlib.jl) (misc functions) and more.
```
Comment (mcabbott) [inline docs/src/index.md:21]:
```suggestion
The **reference** section includes, beside Flux's own functions, those of some companion packages: [Zygote.jl](https://github.com/FluxML/Zygote.jl) (automatic differentiation), [Optimisers.jl](https://github.com/FluxML/Optimisers.jl) (training) and others.
```

File: docs/src/ecosystem.md (hunk 6 above)
Comment (ToucheSir) [reply docs/src/ecosystem.md]: Trying to find a way to express that KNet takes an equally valid but different approach to the ML library design problem. The README and https://denizyuret.github.io/Knet.jl/latest/tutorial/#Philosophy don't give us much in the way of talking points to succinctly contrast against Flux, however.

File: docs/make.jl (hunk 8 above)
Comment (ToucheSir) [reply docs/make.jl:42]: My thought was to take inspiration from the shortened titles, something like so:
```suggestion
            "Low-level Operations - NNlib.jl" => "models/nnlib.md",
            "Taking Gradients - Zygote.jl" => "training/zygote.md",
            "Working with Data - MLUtils.jl" => "data/mlutils.md",
            "Manipulating Models - Functors.jl" => "models/functors.md",
            "One-Hot Encoding - OneHotArrays.jl" => "data/onehot.md",
```

File: docs/src/ecosystem.md (hunk 6 above)
Comment (mcabbott) [reply docs/src/ecosystem.md]: Yes, this seems aimed more at tensorflow et al, parallel to pytorch / Flux / etc.

File: docs/make.jl (hunk 8 above)
Comment (mcabbott) [reply docs/make.jl]: Last commit does this, see what you think (PR before + after):

Screenshot 2022-12-07 at 20 04 48

Comment (ToucheSir) [reply docs/make.jl]: Looks pretty good!

Comment (mcabbott) [reply docs/make.jl]: Shall we do this, then refine as necc?

**Review by ToucheSir — APPROVED** [review]
